Terraform for bigquery retail java application

weix0011 opened the pull request:
**Purpose:** Automate deployment of BigQuery components
**Related files:**
- retail/retail-java-applications/terraform/modules/bigquery/main.tf
- retail/retail-java-applications/terraform/modules/bigquery/variables.tf
- retail/retail-java-applications/terraform/modules/bigquery/store-locations-bq-schema.json

**Purpose:** Organise terraform scripts to modules
**Related files:**
- retail/retail-java-applications/terraform/modules/pubsub/main.tf
- retail/retail-java-applications/terraform/modules/pubsub/variables.tf
- retail/retail-java-applications/terraform/main.tf
- retail/retail-java-applications/terraform/terraform.tfvars
- retail/retail-java-applications/terraform/variables.tf

**Prupose:** Document usage of terraform scripts
**Related files:**
- retail/retail-java-applications/terraform/README.MD

tzehon reviewed (changes requested):
1. If you run terraform init with version 0.14 and above you will get a warning about deprecated provider configuration blocks. Please make the required change to fix the warning

2. The project is currently hardcoded. Please change this as this means the script will not work for *any other project* except your own. The script should be able to work for any user with a GCP project.

weix0011 commented:
1. Added required providers block to fix the warning
2. Changed project id as a variable in variables.tf and terraform.tfvars

tzehon commented:
1. Please always keep your fork up to date with upstream *before* doing anything. It is a best practice to always sync your fork with upstream before you start any work. Think of it as hand washing before having a meal. It will help you avoid merge conflicts in the future. The current repo is 2 commits behind upstream.

2. Squash your commits to keep a linear git history instead of pushing up multiple commits. See:
https://stackoverflow.com/a/15055649/106396
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges#squash-and-merge-your-pull-request-commits
https://cloudfour.com/thinks/squashing-your-pull-requests/
https://git-scm.com/book/en/v2/Git-Branching-Rebasing
These are standard industry best practices and it's good to pick them up early.

3. Please also include a .gitignore that takes into account intermediate Terraform artifacts. See - 
https://git-scm.com/docs/gitignore
https://www.atlassian.com/git/tutorials/saving-changes/gitignore

4. Practice with this current PR. If need be, feel free to close this PR and start a new one.

weix0011 commented:
1. Noted and synced with the upstream repository
2. Tried to rebase on this branch to give a clear message about what this PR is doing
3. Added a .gitignore file, but it is ignored in the parent .gitignore file in line 68 under ignore eclipse files. Shall I just keep it locally? For now I think it is better to edit the parent .gitignore file to include the terraform artifacts for simplicity.

tzehon commented:
Looking good! Now that you're familiar with the workflow and common practices, please refactor the PR in the following ways to make it more production ready:
1. Do not set a default for the project id so the user will have to supply the project id

2. Make this first PR focused on PubSub only. This means the first PR will use Terraform to enable the PubSub API and create the requisite PubSub components. Subsequent components will be in future PRs. Keep each PR focused. standalone and small so it's easy to review and to use. In addition - 

3. Parameterize the names of the PubSub components as variables so the user can choose to supply their own names. For example, not everyone might want to call their topic "Inventory-outbound"

4. Refactor terraform.tfvars by removing the default for project id, related to point 1. Also, you can set the defaults for the PubSub components here. For example, if the user doesn't specify a topic name through --var flags, the default will be "Inventory-outbound", taken from the .tfvars file due to the precedence of variable loading (https://www.terraform.io/docs/language/values/variables.html#variable-definition-precedence).

5. Add a README.md that specifies how to run the Terraform script, similar to the one here: https://github.com/GoogleCloudPlatform/dataflow-sample-applications/tree/master/retail/retail-clickstream-application#deploy-using-terraform

6. Yes, modify the parent .gitignore.

weix0011 commented:
1. Removed the default project ID to prompt the user for input
2. Changed this PR to focus on pubsub related resources only
3. Parameterized pubsub components' names and added a type string constraint as the name only accepts string
4. Refactored terraform.tfvars to include the default values of pubsub components. There is also a default argument when declaring a variable. Can we indicate the default values there, too?
5. Added a README_TERRAFORM.md file as there is an existing README.md file. Is it fine if we have multiple md files?
6. Edited parent .gitignore to exclude terraform intermediate artifacts

tzehon commented:
Regarding 5. -> You can remove the second README and modify the existing README by adding the 'Deploy using Terraform' section, same as here: https://github.com/GoogleCloudPlatform/dataflow-sample-applications/tree/master/retail/retail-clickstream-application#deploy-using-terraform

weix0011 commented:
Updated the existing README file with a terraform section and removed the second README file.

weix0011 committed:
Add terraform scripts for BigQuery and enable related APIs

Organise terraform scripts to modules

tzehon commented on retail/retail-java-applications/terraform/modules/bigquery/main.tf:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright 2020 Google LLC
```
Update year

tzehon commented on retail/retail-java-applications/terraform/modules/bigquery/main.tf:
```diff
@@ -0,0 +1,62 @@
+ * limitations under the License.
+ */
+
+//enable bigquery API
```
Remove comment, the following code block is self explanatory

tzehon commented on retail/retail-java-applications/terraform/modules/bigquery/main.tf:
```diff
@@ -0,0 +1,62 @@
+  disable_on_destroy = false
+}
+
+//create bigquery resources
```
Remove comment, the following code block is self explanatory

tzehon commented on retail/retail-java-applications/terraform/modules/bigquery/variables.tf:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright 2020 Google LLC
```
Update year

tzehon commented on retail/retail-java-applications/terraform/modules/pubsub/main.tf:
```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2020 Google LLC
```
Update year

tzehon commented on retail/retail-java-applications/terraform/modules/pubsub/main.tf:
```diff
@@ -0,0 +1,106 @@
+ * limitations under the License.
+ */
+
+//enable pubsub API
```
Remove comment, the following code block is self explanatory

tzehon commented on retail/retail-java-applications/terraform/modules/pubsub/main.tf:
```diff
@@ -0,0 +1,106 @@
+  disable_on_destroy = false
+}
+
+//create pubsub resources
```
Remove comment, the following code block is self explanatory

tzehon commented on retail/retail-java-applications/terraform/modules/pubsub/variables.tf:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright 2020 Google LLC
```
Update year

tzehon commented on retail/retail-java-applications/terraform/variables.tf:
```diff
@@ -20,34 +20,66 @@ variable "project_id" {
 variable "topic_transactions_inbound" {
-  type    = string
+  type        = string
+  description = "Topic name for transactions inbound"
 }
 
 variable "topic_inventory_inbound" {
-  type    = string
+  type        = string
+  description = "Topic name for inventory inbound"
 }
 
 variable "topic_inventory_outbound" {
-  type    = string
+  type        = string
+  description = "Topic name for inventory outbound"
 }
 
 variable "clickstream_inbound_sub" {
-  type    = string
+  type        = string
+  description = "Subscription for clickstream inbound"
 }
 
 variable "transactions_inbound_sub" {
-  type    = string
+  type        = string
+  description = "Subscription for transactions inbound"
 }
 
 variable "inventory_inbound_sub" {
-  type    = string
```
Do formatting changes in a separate commit from commits containing functional changes. 

Combining formatting changes and functional changes in a commit makes it difficult to isolate what changed, and also makes it difficult to rollback. You can't rollback just the formatting change or the functional change independently, you have to do both together, even though you might want to retain one or the other. This is a best practice.

tzehon reviewed (changes requested):
General comment: Make the refactoring to Terraform modules a separate commit from the BigQuery commit (Same for formatting changes). These are different things that are happening here.

Think of it from another developer's point of view: If I want to look through the history and understand/troubleshoot something, I want to be able to checkout 1 commit that either just does the refactoring to modules, OR a commit that adds BigQuery support. If I have a commit that does both refactoring to modules AND adds BigQuery support, it makes it hard to understand/debug.

tzehon commented on retail/retail-java-applications/terraform/README.MD:
````diff
@@ -114,7 +120,7 @@ google_project_service.pubsub: Creating...
-Apply complete! Resources: 8 added, 0 changed, 0 destroyed.
+Apply complete! Resources: 12 added, 0 changed, 0 destroyed.
 ```
 ## Terraform Destroy
````
Running ```terraform destroy``` gives me this error: ```Error: cannot destroy instance without setting deletion_protection=false and running 'terraform apply'```

weix0011 replied:
- The problem was that an additional flag "deletion_protection" was introduced for big_query_table in [terraform-provider-google](https://github.com/hashicorp/terraform-provider-google/blob/master/CHANGELOG.md) 3.58.0 released on 23 Feb 2021 to prevent accidental deletion of BigQuery tables.
- Previously installed provider version 3.57.0 was reused when running `terraform init` because it satisfied the required provider version `>=3.48.0` and previously used version recorded in .terraform.lock.hcl file.
- The problem was replicated by cloning a new repo and running terraform scripts from there which installed google provider version 3.58.0.
- Will change the required provider version to `>=3.58.0` instead and always test with latest version of provider.
**Question:**
Should we also include .terraform.lock.hcl in the repo as suggested in the [dependency lock file doc](https://www.terraform.io/docs/language/dependency-lock.html)?
